packages/core-utils/src/optimism/batch-encoding.ts

```diff
@@ -9,119 +13,385 @@ export interface BatchContext {
+  // it is serialized
+  public type: BatchType
+
+  constructor(options: Partial<SequencerBatch> = {}) {
+    super()
+    this.contexts = []
+    this.transactions = []
+
+    if (typeof options.shouldStartAtElement === 'number') {
+      this.shouldStartAtElement = options.shouldStartAtElement
+    }
+    if (typeof options.totalElementsToAppend === 'number') {
+      this.totalElementsToAppend = options.totalElementsToAppend
+    }
+    if (Array.isArray(options.contexts)) {
+      this.contexts = options.contexts
+    }
+    if (Array.isArray(options.transactions)) {
+      this.transactions = options.transactions
+    }
+    if (typeof options.type === 'number') {
+      this.type = options.type
+    }
+  }
+
+  write(bw: BufferWriter): BufferWriter {
+    bw.writeBytes(FOUR_BYTE_APPEND_SEQUENCER_BATCH)
+
+    bw.writeU40BE(this.shouldStartAtElement)
+    bw.writeU24BE(this.totalElementsToAppend)
+
+    const contexts = this.contexts.slice()
+    if (this.type === BatchType.ZLIB) {
+      contexts.unshift(
+        new Context({
+          blockNumber: 0,
+          timestamp: 0,
+          numSequencedTransactions: 0,
+          numSubsequentQueueTransactions: 0,
+        })
+      )
+    }
+    bw.writeU24BE(contexts.length)
+
+    for (const context of contexts) {
+      context.write(bw)
+    }
+
+    if (this.type === BatchType.ZLIB) {
+      const writer = new BufferWriter()
+      for (const tx of this.transactions) {
+        tx.write(writer)
+      }
+      const compressed = zlib.deflateSync(writer.render())
+      bw.writeBytes(compressed)
+    } else {
+      // Legacy
+      for (const tx of this.transactions) {
+        tx.write(bw)
+      }
+    }
+
+    return bw
+  }
+
+  read(br: BufferReader): this {
+    const selector = br.readBytes(4)
+    if (Buffer.compare(selector, FOUR_BYTE_APPEND_SEQUENCER_BATCH) !== 0) {
+      br.seek(-4)
+    }
+
+    this.type = BatchType.LEGACY
+    this.shouldStartAtElement = br.readU40BE()
+    this.totalElementsToAppend = br.readU24BE()
+
+    const contexts = br.readU24BE()
+    for (let i = 0; i < contexts; i++) {
+      const context = Context.read<Context>(br)
+      this.contexts.push(context)
+    }
+
+    // handle typed batches
+    if (this.contexts.length > 0 && this.contexts[0].timestamp === 0) {
+      switch (this.contexts[0].blockNumber) {
+        case 0: {
+          this.type = BatchType.ZLIB
+          const bytes = br.readBytes(br.left())
+          const inflated = zlib.inflateSync(bytes)
+          br = new BufferReader(inflated)
+
+          // remove the dummy context
+          this.contexts = this.contexts.slice(1)
+          break
+        }
+      }
+    }
+
+    for (const context of this.contexts) {
+      for (let i = 0; i < context.numSequencedTransactions; i++) {
+        const tx = BatchedTx.read<BatchedTx>(br)
+        this.transactions.push(tx)
+      }
+    }
+
+    return this
+  }
+
+  getSize(): number {
+    if (this.type === BatchType.ZLIB) {
```

Comment:
Since the size isn't know upfront, its not possible to pre-allocate the buffer beforehand. When the size is known, the buffer is pre allocated and then its all written to. When `-1` is returned from `getSize`, every call on the buffer writer will be tracked in an array and then when `encode` is called, it will write all of the data to a buffer